File: java_solutions/241.different-ways-to-add-parentheses.java
/*
 * @lc app=leetcode id=241 lang=java
 *
 * [241] Different Ways to Add Parentheses
 */

class Solution {
    public java.util.List<Integer> diffWaysToCompute(String expression) {
        return java.util.Optional.of(java.util.stream.IntStream.range(0, expression.length()).filter(i -> !Character.isDigit(expression.charAt(i))).boxed().flatMap(i -> diffWaysToCompute(expression.substring(0, i)).stream().flatMap(l -> diffWaysToCompute(expression.substring(i + 1)).stream().map(r -> expression.charAt(i) == '+' ? l + r : expression.charAt(i) == '-' ? l - r : l * r))).collect(java.util.stream.Collectors.toList())).map(res -> res.isEmpty() ? java.util.Collections.singletonList(Integer.parseInt(expression)) : res).get();
    }
}

File: kotlin_solutions/241.different-ways-to-add-parentheses.kt
/*
 * @lc app=leetcode id=241 lang=kotlin
 *
 * [241] Different Ways to Add Parentheses
 */

class Solution {companion object { private val memo = mutableMapOf<String, List<Int>>() }fun diffWaysToCompute(expression: String): List<Int> = memo.getOrPut(expression) { expression.indices.filter { !expression[it].isDigit() }.flatMap { i -> diffWaysToCompute(expression.substring(0, i)).flatMap { l -> diffWaysToCompute(expression.substring(i + 1)).map { r -> if (expression[i] == '+') l + r else if (expression[i] == '-') l - r else l * r } } }.ifEmpty { listOf(expression.toInt()) } }}
